File: ruby/simple_cipher.rb
class Cipher
  attr_reader :key, :alpha, :key_codes

  def initialize(key_input=nil)
    @alpha=("a".."z").to_a * 2
    @key=generate_key(key_input)
    @key_codes=key.split("").map{|char| alpha.index(char)}
  end

  def encode(phrase)
    manipulate_text(phrase, encoding=true)
  end

  def decode(phrase)
    manipulate_text(phrase, encoding=false)
  end

  def manipulate_text(phrase, output="", encoding)
    phrase.split("").map{|char| alpha.index(char)}.each_with_index do |index1, index2|
      encoding ? index1+=key_codes[index2] : index1-=key_codes[index2]
      output+=alpha[index1]
    end
    output
  end

  def generate_key(key_input)
    key_input ||= alpha.sample*100
    raise ArgumentError, "Invalid Key" if key_input.match(/[^a-z]/) || key_input == ""
    key_input
  end
end

File: js/simple-cipher.js
Cipher=function(key){
  this.key = checkValidity(key)
  this.encoding = getPositions(this.key);
};

function checkValidity(key){
  if (key==undefined){return 'bbbbbbbbbb'}
  else if (key.toUpperCase()==key){throw("Bad key")}
  else {return key};
};

function getPositions(text){
  return text.split("").map(function(letter){return alpha.indexOf(letter)})
}

Cipher.prototype.encode = function(text){
  var original = getPositions(text);
  var encoding = this.encoding;
  var transformed = original.map(function(value, index){
    return alpha[(value+encoding[index])%26]
  });
  return transformed.join("")
};

Cipher.prototype.decode = function(text){
  var original = getPositions(text);
  var encoding = this.encoding;
  var transformed = original.map(function(value, index){
    return alpha[(value-encoding[index]+26)%26]
  });
  return transformed.join("")
};

var alpha = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm',
  'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z']

module.exports = Cipher;
